include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,482 @@
+     * If the current wrapper vector has ghost elements,
+     * then <code> VectorType::operator=(const VectorType&) </code> should still
+     * be allowed on it.
+     */
+    void set (const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Perform addition.
+     */
+    void plus (const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Scale the wrapper vector by @p alpha and add ROL::Vector @p rol_vector
+     * to it.
+     */
+    void axpy (const value_type               alpha,
+               const ROL::Vector<value_type> &rol_vector);
+
+    /**
+     * Scale the wrapper vector.
+     */
+    void scale (const value_type alpha);
+
+    /**
+     * Return the dot product with a given ROL::Vector @p rol_vector.
+     */
+    value_type dot (const ROL::Vector<value_type> &rol_vector) const;
+
+    /**
+     * Return the \f$ L^{2} \f$ norm of the wrapped vector.
+     *
+     * The returned type is of VectorAdaptor::value_type so as to maintain
+     * consistency with ROL::Vector<VectorAdaptor::value_type> and
+     * more importantly to not to create an overloaded version namely,
+     * <code> VectorAdaptor::real_type norm() const; </code>
+     * if real_type and value_type are not of the same type.
+     */
+    value_type norm() const;
+
+    /**
+     * Return a clone of the wrapped vector.
+     */
+    Teuchos::RCP<ROL::Vector<value_type>> clone() const;
+
+    /**
+     * Create and return a Teuchos smart reference counting pointer to the basis
+     * vector corresponding to the @p i \f${}^{th}\f$ element of
+     * the wrapper vector.
+     */
+    Teuchos::RCP<ROL::Vector<value_type>> basis (const int i) const;
+
+    /**
+     * Apply unary function @p f to all the elements of the wrapped vector.
+     */
+    void applyUnary (const ROL::Elementwise::UnaryFunction<value_type> &f);
+
+    /**
+     * Apply binary function @p f along with ROL::Vector @p rol_vector to all
+     * the elements of the wrapped vector.
+     */
+    void applyBinary (const ROL::Elementwise::UnaryFunction<value_type> &f,
+                      const ROL::Vector<value_type>                     &rol_vector);
+
+    /**
+     * Return the accumulated value on applying reduction operation @p r on
+     * all the elements of the wrapped vector.
+     */
+    value_type reduce (const ROL::Elementwise::ReductionOp<value_type> &r) const;
+
+    /**
+     * Print the wrapped vector to the output stream @p outStream.
+     */
+    void print (std::ostream &outStream) const;
+
+  };
+
+
+  /*------------------------------member definitions--------------------------*/
+#ifndef DOXYGEN
+
+
+  template<typename VectorType>
+  VectorAdaptor<VectorType>::
+  VectorAdaptor (const Teuchos::RCP<VectorType> &vector_ptr)
+    :
+    vector_ptr (vector_ptr)
+  {
+    static_assert (std::is_convertible<real_type, value_type>::value,
+                   "The real_type of the current VectorType is not "
+                   "convertible to the value_type.");
```

Comment:
I think it would be easier if you put this `static_assert` into the class declaration itself. Any attempt to declare such a vector adaptor is already doomed, not just the creation, so getting the error earlier rather than later seems like the right idea.